cmd/slicetrace/session.go

```diff
@@ -0,0 +1,262 @@
+		}
+		invByIndex[index] = invocation{
+			index:    index,
+			location: event.Args["location"].(string),
+			args:     truncatef(strings.Join(args, ", ")),
+		}
+	}
+	invs := make([]invocation, 0, len(invByIndex))
+	for _, inv := range invByIndex {
+		invs = append(invs, inv)
+	}
+	sort.Slice(invs, func(i, j int) bool { return invs[i].index < invs[j].index })
+	return invs
+}
+
+func buildTasks(events []trace.Event) []task {
+	var tasks []task
+	for _, event := range events {
+		if event.Cat != "task" {
+			continue
+		}
+		matches := reTask.FindStringSubmatch(event.Name)
+		if matches == nil {
+			log.Printf("could not parse name: %#v", event)
+			continue
+		}
+		invIndex, err := strconv.Atoi(matches[1])
+		if err != nil {
+			panic("should be digits")
+		}
+		op := matches[2]
+		shard, err := strconv.Atoi(matches[3])
+		if err != nil {
+			panic("should be digits")
+		}
+		readDuration := int64(event.Args["readDuration"].(float64))
+		writeDuration := int64(event.Args["writeDuration"].(float64))
+		tasks = append(tasks, task{
+			invIndex: invIndex,
+			op:       op,
+			shard:    shard,
+			start:    time.Duration(event.Ts * 1e3),
+			duration: time.Duration(event.Dur * 1e3),
+			read:     time.Duration(readDuration * 1e3),
+			write:    time.Duration(writeDuration * 1e3),
+		})
+	}
+	return tasks
+}
+
+func buildOpStats(tasks []task) []opStat {
+	type invOp struct {
+		invIndex int
+		op       string
+	}
+	type accum struct {
+		maxShard  int
+		minStart  time.Duration
+		maxEnd    time.Duration
+		durations []time.Duration
+		total     time.Duration
+		read      time.Duration
+		write     time.Duration
+	}
+	accums := make(map[invOp]*accum)
+	for _, task := range tasks {
+		invOp := invOp{task.invIndex, task.op}
+		a, ok := accums[invOp]
+		if !ok {
+			a = &accum{
+				minStart: 1<<63 - 1,
+				// Otherwise zero values are fine.
+			}
+			accums[invOp] = a
+		}
+		if task.start < a.minStart {
+			a.minStart = task.start
+		}
+		end := task.start + task.duration
+		if a.maxEnd < end {
+			a.maxEnd = end
+		}
+		if a.maxShard < task.shard {
+			a.maxShard = task.shard
+		}
+		a.durations = append(a.durations, task.duration)
+		a.total += task.duration
+		a.read += task.read
+		a.write += task.write
+	}
+	opStats := make([]opStat, 0, len(accums))
+	for invOp, a := range accums {
+		sort.Slice(a.durations, func(i, j int) bool {
+			return a.durations[i] < a.durations[j]
+		})
+		q1, q2, q3 := computeQuartiles(a.durations)
+		opStats = append(opStats, opStat{
+			invIndex: invOp.invIndex,
+			op:       invOp.op,
+			shards:   a.maxShard + 1,
+			start:    a.minStart,
+			duration: a.maxEnd - a.minStart,
+			total:    a.total,
+			read:     a.read,
+			write:    a.write,
+			min:      a.durations[0],
+			q1:       q1,
+			q2:       q2,
+			q3:       q3,
+			max:      a.durations[len(a.durations)-1],
+		})
+	}
+	return opStats
+}
+
+func computeQuartiles(ds []time.Duration) (time.Duration, time.Duration, time.Duration) {
```

Comment:
It's not very likely, but it'd be nice to not panic for [any number of tasks](https://play.golang.org/p/5nu5fh76Tdv).